### ISSUE TITLE
R1 calibration

### ISSUE THREAD

**Issue by dneise** [cta-sst-1m/digicampipe#155]
Can anyone tell me where these numbers come from?

https://github.com/cta-sst-1m/digicampipe/blob/acd9fa0c3bc660d4c82cee7c511009cf2631027f/digicampipe/utils/calib.py#L4-L30

So far I only found out that:

    nsb_rate = np.logspace(-3, np.log10(50), 30)

The other two, do also not really look like measurements, but more like coarsly samples functions.

I must admit in the beginning when I saw these numbers, I assume they were measurements .. but after plotting them, I am not so sure anymore.

**Comment by calispac** [comment]
Yes I have to admit it seems coming from nowhere.
Basically it is a lookup table to relate: 
- pedestal_shift to NSB rate
- pedestal_sdt to NSB rate

They were obtained with `digicamtoy`. This parametrization was verified with data comparison.
However now we have an analytical description (to be added in software soon).

**Comment by dneise** [comment]
> However now we have an analytical description (to be added in software soon).

Cool I was not aware, could you give me the link?

**Comment by calispac** [comment]
It is part of the calibration document... :D sorry...

**Comment by dneise** [comment]
> They were obtained with digicamtoy. This parametrization was verified with data comparison.

Erm ... I must have made a mistake .. I tried to plot these 4 functions ... but two of them do not really look correct to me.

![image](https://user-images.githubusercontent.com/8200858/37159329-d62d0408-22ed-11e8-8e07-d8d0bfee466a.png)

The plottling code is this:
```python
import numpy as np
import matplotlib.pyplot as plt
%matplotlib inline

from digicampipe.utils import calib

baseline_shift = np.linspace(calib.baseline_shift.min(), calib.baseline_shift.max(), 100)
std = np.linspace(calib.std.min(), calib.std.max(), 100)

fig, ((upper_left, upper_right),(lower_left, lower_right)) = plt.subplots(2, 2, figsize=(16, 8))
upper_left.plot(
    baseline_shift,
    calib.compute_gain_drop(baseline_shift, 'mean'),
    '.:'
)
upper_left.set_xlabel("baseline shift [a.u.]")
upper_left.set_ylabel("gain_drop [a.u.]")
upper_left.grid()

upper_right.plot(
    std,
    calib.compute_gain_drop(std, 'std'),
    '.:'
)
upper_right.set_xlabel("baseline std-dev[a.u.]")
upper_right.set_ylabel("gain_drop [a.u.]")
upper_right.grid()

lower_left.plot(
    baseline_shift,
    calib.compute_nsb_rate(baseline_shift, 'mean'),
    '.:'
)
lower_left.set_xlabel("baseline shift [a.u.]")
lower_left.set_ylabel("nsb rate [a.u.]")
lower_left.grid()

lower_right.plot(
    std,
    calib.compute_nsb_rate(std, 'std'),
    '.:'
)
lower_right.set_xlabel("baseline std-dev[a.u.]")
lower_right.set_ylabel("nsb rate [a.u.]")
lower_right.grid()
```

**Comment by calispac** [comment]
Seems like the problem comes from the parametrization with pedestal_std only. In particular from : `    calib.compute_nsb_rate(std, 'std')`.
 Now I remember the issue it is because the function : 

std(nsb_rate) as a local maximum (not a bijection) : 
![nsb_rate_vs_std](https://user-images.githubusercontent.com/23462345/37242813-b46ee088-246f-11e8-8419-8f330795de65.png)

**Comment by dneise** [comment]
Yes that is exactly the source of the problem. Sorry, I should have also shown this plot in the picture above... I had it in that notebook, but I did not show it here, since it did not show the problem as nicely .. so I deleted it.

You say:
> The fact that std is not strictly increasing with nsb rate seems wrong to me.
And I fully agree here .. therefore my initial question was: "Can anyone tell me where these numbers come from?"  So that we could further investigate the source of this local maximum... 

I think it is of high importance to quote the source of (magic) numbers when used inside of code. Or when read in as a look-up-table or so. This goes into the direction of provenance... numbers in the code-base which are not traceable do not break reproducibility, so having the numbers in the code, without knowing where they come from is much better than e.g. reading them in from a text-file and the text-file is missing in the repo ... 

But still not referencing the (internal) paper these numbers came from makes it hard to understand. In this case, you said, these numbers came from a simulation. So I assume this simulation was checked in somewhere and could be referenced here. like this:
```
# These numbers come from a simulation with digicamtoy, details can be found here:
# https://github.com/cta-sst-1m/digicamtoy/blob/master/gain_drop_statistic.py
```

Of course this still has the problem that some random guys named dominik, might have a look at that script and say: "the name is bad, let's rename it" ... not knowing that this script is an important reference... one solution is to always use perma-links to cite simulation-code, which yielded a certain result which is the basis for further decisions. On github pages, I think one can still hit the `y`-key on the keyboard and get a perma-link to that particular point in time of the repo, like:
```
https://github.com/cta-sst-1m/digicamtoy/blob/8c5c3df1734375ac9bba9bad35a59c9a17bb765d/gain_drop_statistic.py
```

-----

Another point which is interesting I think:

> For the moment I suggest we forget about evaluating nsb/gain_drop with std. Anyhow we've never used it

A reader of the code might spend hours understanding code ... which then later turns out to have never been used at all. Maybe somebody produced that code in 30 minutes ... planning to use it later  ... then something else was suddenly more important and that code was forgotten. I totally understand, but still is might be great time-eating-monster. 

I personally assumed that using the baseline-std-deviation is much more often used, since one is independent of a dark-baseline run, which might or might not exists...

**Comment by dneise** [comment]
See for example a first try to get some of our code into ctapipe. 
https://github.com/dneise/ctapipe/blob/bbd53c5981d4400e34a06144588814e24547f809/ctapipe/calib/camera/r1.py#L426-L436

There is this `if dark-baseline do this, else do that` .. and now I learn that in reality only one branch of this was ever used.

**Comment by dneise** [comment]
I used the digigcamtoy to create the same plot again. I varied the gain an non-surprisingly, when the gain is low enough that the gain-drop stays low even at elevated rates (find quantitative expression), the std-deviation of the baselineshift versus the nsb-rate stays nicely bijective.

![baseline_std_vs_nsb_rate_and_gain](https://user-images.githubusercontent.com/8200858/37388354-69aa3e96-2760-11e8-951a-3d7445c9686d.png)

In the digicamtoy the default value for the gain is close to 6 ... in digicampipe I've seen 23... So I wanted to ask, if there is any source to learn what typical gain values in digicam are?

**Comment by calispac** [comment]
Thanks for the plots! Ok not sure yet I fully understand the meaning of the result you show. As before it seems puzzeling... 
They are the same.
1. Nominal gain is ~ 5.8 ADC/p.e. (a pulse of amplitude 5.8 ADC corresponds to a 1 p.e. pulse).
2. Nominal "intagrated gain" is ~ 23 ADC/p.e. ( if the area of a pulse is 23 ADC that corresponds to a 1 p.e. pulse)

Now I am considering in the calibration I am normalizing pulse integrals such that the area of the integrated pulse is equal to the amplitude of the regular pulse.

**Comment by calispac** [comment]
Also could you push the plot you have to ``digicamtoy` in examples ? either as a regular python file or as a notebook.

**Comment by dneise** [comment]
Ah I see ... So both numbers are correct.. of course ... damn not enough sleep.

I think, when 23 is the result of an integration then its unit might be (LSB * samples) or (LSB * ns)... or so.

Sure .. erm .. I'll open a new PR for that plot only so it does not get mixed up with the other pep8 stuff, I was playing around with

**Comment by calispac** [comment]
> I think, when 23 is the result of an integration then its unit might be (LSB * samples) or (LSB * ns)... or so.

yes you are right, less confusing. 
But anyway LSB is an arbitrary unit so one can call everything a LSB. For instance if before digitization you have a component that takes the signal and returns the integrated signal over a certain window. You would still call this output a LSB

**Comment by dneise** [comment]
Yes that is correct, and indeed all ADCs do a bit of integrating/averaging while their "sample-and-hold" circuitry is sampling. 

In FACT we did some calibration of the LSBs to actual voltage, measured at some point after the pre-amp. However we never bothered to really calibrate this voltage signal to the current pulse flowing through the SiPM... so it was pretty worthless. LSB, mV ... in principle it could have been also "apples" ... in the end all what counts is what the gain is.

Anyway when we did some integration of this voltage over a range of time, of course the result had some unit like "mV * ns" or "mV * samples" depending on how the guy doing it liked to work with time  ... a sample in FACT is more or less exactly 0.5ns .. so when people did not explicitely state the unit, there was always a factor 2 to be missed easily. I guess in digicam where a sample is 4ns .. it is very easy to miss this factor of 4... so stating the units explicitely might help, once somebody defined what you in digicam mean with an LSB, everything is defined and not arbitrary anmore...

However, one more comment ... Just stating that the gain is 23 LSB * ns when integrating over the pulse might also miss a little bit of important information. The question is often how you defined your integration window, and how did you define the baseline? 

Otherwise two people finding pulses and integrating over them, might do it differently and come up with totally different numbers of p.e. when using the same number for the gain.

I assume 23 is the result when integrating the entire pulse over dunno 100ns using trapz? While in digicampipe I've seen one is integrating over 7 slices using simple-summing-rule.

**Comment by calispac** [comment]
> The question is often how you defined your integration window, and how did you define the baseline?

Yes totally agree that's why a p.e. is not : my adc / gain but rather: 

my computed charge with method X / the gain I obtained with method X

> Otherwise two people finding pulses and integrating over them, might do it differently and come up with totally different numbers of p.e. when using the same number for the gain.

That's why I want to normalize each charge to the same "units"

> I assume 23 is the result when integrating the entire pulse over dunno 100ns using trapz? While in digicampipe I've seen one is integrating over 7 slices using simple-summing-rule.

Thanks for asking. It is 7.

**Comment by calispac** [comment]
So doing some calculations I found out that there is a maximum at (which is what your plot tends to show): 

f_{nsb} = 1/(RC) = 1.17 GHz (for all gains)

This is only due because of gain drop!

**Comment by dneise** [comment]
cool can you show the calculations? (a picture of the whiteboard / piece of paper is fine if you got no time to digitize it )
